### ISSUE TITLE
clearTimeout() on componentWillUnmount

### ISSUE THREAD

**Issue by ricardojrgpimentel** [fkhadra/react-toastify#117]
I'm getting this error when i navigate to another page:

```
Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the ToastContainer component.
```

Shouldn't the component clearTimeout() when it unmounts?

**Comment by fkhadra** [comment]
Hello @ricardojrgpimentel ,

Could you tell me which version you are using?

Thanks.

**Comment by ricardojrgpimentel** [comment]
@fkhadra I'm on 3.2.1

**Comment by fkhadra** [comment]
Actually, the eventManager and react-transition-group use `setTimeout`. 
Could you provide a snippet so I can reproduce the error? 

I'll try to reproduce the issue on my side also.

**Comment by ricardojrgpimentel** [comment]
I'm using react-router, and i have the `ToastContainer` inside `ComponentA` and `ComponentB`

When the route changes to something like `/site/componentA` the `ComponentA` gets mounted with `ToastContainer` along with a `toast` that lasts for 10 seconds, if i navigate to `/site/componentB` the same happens but this time i get the error stated above.

**Comment by fkhadra** [comment]
Thanks for the explanation. I'll investigate.

**Comment by ricardojrgpimentel** [comment]
It's active

**Comment by fkhadra** [comment]
@ricardojrgpimentel I'm trying to reproduce the issue here:
[![Edit o4xkk1pq9y](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/o4xkk1pq9y)

Could you modify the snippet to fit your use case please ?

**Comment by realneeraj** [comment]
@fkhadra 
I have updated the code page to reproduce the error. Please check the Home Component and try to login. Also check console log for the error.

Please let me know the solution.

Thanks in advance.

**Comment by fkhadra** [comment]
@realneeraj could you share the link? The sandbox is still showing the old code.

Thanks

**Comment by realneeraj** [comment]
@fkhadra 
https://codesandbox.io/s/64mml459z - Hopefully you can see it here now?

-------------------------------------------- Source code is here --------------------------------------------
import React from "react";
import { render } from "react-dom";
import {
  BrowserRouter as Router,
  Route,
  Link,
  Redirect
} from "react-router-dom";
import { ToastContainer, toast } from "react-toastify";

class Home extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      redirect: false
    }
  }
  componentDidMount() {
    toast("HOME");
  }
  notify = () => {
    toast.success("Logged in successfully !")
    this.setState({
      redirect: true
    })
  };

  render() {
    if(this.state.redirect){
      return <Redirect to={'/about'} />
    }
    return (
      <div>
        <h2>Home</h2>
        <ToastContainer />
        <button onClick={this.notify}>Login !</button>
      </div>
    );
  }
}

class About extends React.Component {
  componentDidMount() {
    toast("ABOUT");
  }

  render() {
    return (
      <div>
        <h2>Home</h2>
        <ToastContainer />
      </div>
    );
  }
}

const App = () => (
  <Router>
    <div>
      <button onClick={() => toast("hello")}>TOAST</button>
      <ul>
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
          <Link to="/about">About</Link>
        </li>
      </ul>

      <hr />

      <Route exact path="/" component={Home} />
      <Route path="/about" component={About} />
    </div>
  </Router>
);

render(<App />, document.getElementById("root"));

-------------------------------------------------------------------------------
Basically I would do a login and if its success I would want to show a success toast and redirect the user to inner page.

Hope it makes sense!

**Comment by fkhadra** [comment]
The link is good now. Thanks

**Comment by fkhadra** [comment]
@realneeraj `toast.success("Logged in successfully !")` will never works.
 
When the toast function is called, the `ToastContainer` will call `setState`, in the meantime you are calling `setState({ redirect: true })` which unmount the ToastContainer. 

This is why we get:
> Warning: Can only update a mounted or mounting component. This usually means you called setState, replaceState, or forceUpdate on an unmounted component. This is a no-op.
Please check the code for the ToastContainer component.

Instead of putting `ToastContainer` inside each route component, put it on the App level like:

```js
const App = () => (
  <Router>
    <div>
      <button onClick={() => toast("hello")}>TOAST</button>
      <ul>
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
          <Link to="/about">About</Link>
        </li>
      </ul>
      <hr />
      <Route exact path="/" component={Home} />
      <Route path="/about" component={About} />
      <ToastContainer />
    </div>
  </Router>
);
```

**Comment by realneeraj** [comment]
Thanks for being a great help @fkhadra 

That solution makes more sense!

**Comment by fkhadra** [comment]
You are welcome @realneeraj. @ricardojrgpimentel you were encountering that error probably for the same reason. I'll close the issue. Feel free to create a new one is the issue is not related.

**Comment by ricardojrgpimentel** [comment]
Sorry for the late answer, i moved the ToastContainer to the router component parent to fix it, like you said in the last comment.
